src/main/java/au/org/emii/ncdfgenerator/NcdfEncoderBuilder.java

```diff
@@ -0,0 +1,1476 @@
+		if(!m.find())
+		{
+			throw new NcdfGeneratorException( "couldn't parse attribute date");
+		}
+		unit = m.group(1);
+		String epochString = m.group(2);
+		try {
+			SimpleDateFormat df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss Z");
+			Date ts = df.parse(epochString);
+			epoch = (Long) ts.getTime() / 1000 ;
+		} catch( Exception e )
+		{
+			throw new NcdfGeneratorException( "couldn't extract timestamp '" + epochString + "' " + e.getMessage()  );
+		}
+
+		fill = Float.valueOf( attributes.get( "_FillValue" )).floatValue();
+	}
+
+	public void encode( Array array, int ima,  Object value ) throws NcdfGeneratorException
+	{
+
+		if( value == null) {
+			array.setFloat( ima, fill );
+		}
+		else if( value instanceof java.sql.Timestamp ) {
+			long seconds =  ((java.sql.Timestamp)value).getTime() / 1000  ;
+			long ret = seconds - epoch;
+			if( unit.equals("days"))
+				ret /= 86400;
+			else if( unit.equals("minutes"))
+				ret /= 1440;
+			else if ( unit.equals("seconds"))
+				;
+			else
+				throw new NcdfGeneratorException( "unrecognized time unit " + unit );
+
+			array.setFloat( ima, (float) ret );
+		}
+		else {
+			throw new NcdfGeneratorException( "Not a timestamp" );
+		}
+	}
+}
+
+
+class FloatValueEncoder implements IValueEncoder
+{
+	FloatValueEncoder()
+	{
+		this.fill = 1234;
+	}
+
+	float fill;
+
+	public DataType targetType()
+	{
+		return DataType.FLOAT;
+	}
+
+	public void prepare( Map<String, String> attributes )
+	{
+		fill = Float.valueOf( attributes.get( "_FillValue" )).floatValue();
+
+	}
+
+	public void encode( Array array, int ima, Object value ) throws NcdfGeneratorException
+	{
+		if( value == null) {
+			array.setFloat( ima, fill );
+		}
+		else if( value instanceof Float ) {
+			array.setFloat( ima, (Float) value);
+		}
+		else if( value instanceof Double ) {
+			array.setFloat( ima, (float)(double)(Double) value);
+		}
+		else {
+			throw new NcdfGeneratorException( "Failed to coerce type '" + value.getClass() + "' to float" );
+		}
+	}
+}
+
+
+class IntValueEncoder implements IValueEncoder
+{
+	// Int is 32bit in Netcdf
+
+	IntValueEncoder()
+	{
+		this.fill = 1234;
+	}
+
+	int fill;
+
+	public DataType targetType()
+	{
+		return DataType.INT;
+	}
+
+	public void prepare( Map<String, String> attributes )
+	{
+		fill = Integer.valueOf( attributes.get( "_FillValue" )).intValue();
+	}
+
+	public void encode( Array array, int ima, Object value ) throws NcdfGeneratorException
+	{
+		if( value == null) {
+			array.setInt( ima, fill );
+		}
+		else if( value instanceof Integer ) {
+			array.setInt( ima, (Integer) value);
+		}
+		else if( value instanceof Long ) {
+			array.setInt( ima, (int)(long)(Long) value);
+		}
+		else {
+			throw new NcdfGeneratorException( "Failed to coerce type '" + value.getClass() + "' to float" );
+		}
+	}
+}
+
+
+class ByteValueEncoder implements IValueEncoder
+{
+	ByteValueEncoder()
+	{
+		this.fill = 0x0;
```

Comment:
I agree this is a bit messy. If java had a good Option type, then it would be preferable to use that, and initialize it with a Nothing value, then set it later on the prepare call.

I considered using a boxed Java value and using null to encode the non-initialized state. There are some advantages to this, but in the end, I felt it would just complicate it more than clarify what's going on.  So basically the fill value is just initialized with a dummy value following the principle that all vars should be initialized on creation.